src/Microsoft.SqlTools.ServiceLayer/Connection/ListDatabaseRequestHandler.cs

```diff
@@ -4,223 +4,177 @@
-    )
-SELECT db.name, state_desc AS state, size
-FROM sys.databases db LEFT JOIN db_size ON db.name = db_size.name
-WHERE state_desc='ONLINE'
-ORDER BY name ASC
-";
-            }
-        }
-    }
+	public class ListDatabasesRequestDatabaseProperties
+	{
+		public const string Name = "name";
+		public const string SizeInMB = "sizeInMB";
+		public const string State = "state";
+		public const string LastBackup = "lastBackup";
+	}
+
+	/// <summary>
+	/// Factory class for ListDatabasesRequest handler
+	/// </summary>
+	static class ListDatabaseRequestHandlerFactory
+	{
+		public static IListDatabaseRequestHandler getHandler(bool includeDetails, bool isSqlDB)
+		{
+			if (!includeDetails)
+			{
+				return new DatabaseNamesHandler();
+			}
+			else if (isSqlDB)
+			{
+				return new SqlDBDatabaseDetailHandler();
+			}
+			else
+			{
+				return new SqlServerDatabaseDetailHandler();
+			}
+		}
+	}
+
+	/// <summary>
+	/// Interface of ListDatabasesRequest handler
+	/// </summary>
+	interface IListDatabaseRequestHandler
+	{
+		ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo);
+	}
+
+	/// <summary>
+	/// Base handler
+	/// </summary>
+	abstract class ListDatabaseRequestHandler<T> : IListDatabaseRequestHandler
+	{
+		private static readonly string[] SystemDatabases = new string[] { "master", "model", "msdb", "tempdb" };
+
+		public ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo)
+		{
+			ServerConnection serverConnection = ConnectionService.OpenServerConnection(connectionInfo);
+			SMO.Server server = new SMO.Server(serverConnection);
+
+			var results = new List<T>();
+			ListDatabasesResponse response = new ListDatabasesResponse();
+
+			// Using this function to set the default init properties of the server so that those things can be fetched in a single query.
+			setDefaultServerInitProperties(server);
+
+			foreach (SMO.Database db in server.Databases)
+			{
+				results.Add(CreateItem(db));
+			}
+
+			results = results.Where(s => SystemDatabases.Any(x => this.NameMatches(x, s))).Concat(
+						 results.Where(s => SystemDatabases.All(x => !this.NameMatches(x, s)))).ToList();
+
+			SetResponse(response, results.ToArray());
+			serverConnection.Disconnect();
+
+			return response;
+		}
+
+		protected abstract void setDefaultServerInitProperties(SMO.Server server);
+		protected abstract bool NameMatches(string databaseName, T item);
+		protected abstract T CreateItem(SMO.Database database);
+		protected abstract void SetResponse(ListDatabasesResponse response, T[] results);
+	}
+
+	/// <summary>
+	/// database names handler
+	/// </summary>
+	class DatabaseNamesHandler : ListDatabaseRequestHandler<string>
+	{
+		protected override void setDefaultServerInitProperties(SMO.Server server)
+		{
+			server.SetDefaultInitFields(typeof(Database), new[] { nameof(Database.Name) });
```

Comment:
Did you test this? This shouldn't be necessary - names are one of the properties that are always fetched (since the server needs to know about them to populate the collection)

Reply:
Yeah, the names are fetched in a single query. I'll remove this